x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/PhysicalPlanOptimizer.java

```diff
@@ -123,13 +128,21 @@ public PhysicalPlan apply(PhysicalPlan plan) {
                     // otherwise expect a Fragment
                     if (child instanceof FragmentExec fragmentExec) {
                         var logicalFragment = fragmentExec.fragment();
+                        var skipProject = false;
                         // no need for projection when dealing with aggs
                         if (logicalFragment instanceof Aggregate) {
-                            attributes.clear();
+                            skipProject = true;
                         }
                         var selectAll = projectAll.get();
-                        if (attributes.isEmpty() == false || selectAll) {
+                        if (skipProject == false) {
```

Comment:
This is the gist of the fix - previously the case was detected by removing all the attributes and thinking that if there are no attributes, there's a projection in place - which clearly is not the case.
Hence why the fix is to introduce a separate flag and actually consider the list independently.

Reply:
nit: why not skip `skipProject` and just move the block inside the `if (logicalFragment instanceof Aggregate) {` branch?